pallets/parachain-staking/src/tests.rs

```diff
@@ -2516,6 +2516,67 @@ fn unlock_unstaked() {
 		});
 }
 
+#[test]
+fn kick_candidate_with_full_unstaking() {
+	ExtBuilder::default()
+		.with_balances(vec![(1, 200), (2, 200), (3, 200)])
+		.with_collators(vec![(1, 200), (2, 200), (3, 200)])
+		.build()
+		.execute_with(|| {
+			let max_unstake_reqs: usize = <Test as Config>::MaxUnstakeRequests::get()
+				.saturating_sub(1)
+				.saturated_into();
+			// Fill unstake requests
+			for block in 1u64..1u64.saturating_add(max_unstake_reqs as u64) {
+				System::set_block_number(block);
+				assert_ok!(StakePallet::candidate_stake_less(Origin::signed(3), 1));
+			}
+			assert_eq!(StakePallet::unstaking(3).into_inner().len(), max_unstake_reqs);
+
+			// Additional unstake should fail
+			System::set_block_number(100);
+			assert_noop!(
+				StakePallet::candidate_stake_less(Origin::signed(3), 1),
+				Error::<Test>::NoMoreUnstaking
+			);
+
+			// Fill last unstake request by removing candidate and unstaking all stake
+			assert_ok!(StakePallet::force_remove_candidate(Origin::root(), 3));
+			assert_eq!(StakePallet::unstaking(3).into_inner().len(), max_unstake_reqs + 1);
+		});
+}
+#[test]
+fn kick_delegator_with_full_unstaking() {
+	ExtBuilder::default()
+		.with_balances(vec![(1, 200), (2, 200), (3, 200), (4, 200), (5, 200), (6, 200)])
+		.with_collators(vec![(1, 200)])
+		.with_delegators(vec![(2, 1, 200), (3, 1, 200), (4, 1, 200), (5, 1, 200)])
+		.build()
+		.execute_with(|| {
+			let max_unstake_reqs: usize = <Test as Config>::MaxUnstakeRequests::get()
+				.saturating_sub(1)
+				.saturated_into();
+			// Fill unstake requests
+			for block in 1u64..1u64.saturating_add(max_unstake_reqs as u64) {
+				System::set_block_number(block);
+				assert_ok!(StakePallet::delegator_stake_less(Origin::signed(5), 1, 1));
+			}
+			assert_eq!(StakePallet::unstaking(5).into_inner().len(), max_unstake_reqs);
+
+			// Additional unstake should fail
+			System::set_block_number(100);
+			assert_noop!(
+				StakePallet::delegator_stake_less(Origin::signed(5), 1, 1),
+				Error::<Test>::NoMoreUnstaking
+			);
+
+			// Fill last unstake request by replacing delegator
+			assert_ok!(StakePallet::join_delegators(Origin::signed(6), 1, 200));
+			assert_eq!(StakePallet::unstaking(5).into_inner().len(), max_unstake_reqs + 1);
+			assert!(!StakePallet::is_delegator(&5));
+		});
```

Comment:
The last assert will fail. Maybe we need to add a check to `join_delegators` that at least 1 free unstaking slot is left?

```suggestion
			// Rejoin and kick again
			assert_ok!(StakePallet::join_delegators(Origin::signed(5), 1, 201));
			assert!(StakePallet::is_delegator(&5));
			assert_ok!(StakePallet::join_delegators(Origin::signed(6), 1, 202));
			assert!(!StakePallet::is_delegator(&5));
		});
```

Reply:
> The last assert will fail. Maybe we need to add a check to `join_delegators` that at least 1 free unstaking slot is left?

Yes I was also thinking of this.

Reply:
Please check 19b7c61